### PR TITLE
[aotinductor] Add a demo tutorial

### PR DESCRIPTION
**Pull request by desertfire**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112457

cc @svekars @carljparker

### CODE REVIEW

File: docs/source/torch.compiler_aot_inductor.rst [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    subject to backwards compatibility breaking changes.
+
+AOTInductor is a specialized version of
+`TorchInductor <https://dev-discuss.pytorch.org/t/torchinductor-a-pytorch-native-compiler-with-define-by-run-ir-and-symbolic-shapes/747>`__
+, designed to process exported PyTorch models, optimize them, and produce shared libraries as well
+as other relevant artifacts.
+These compiled artifacts are specifically crafted for deployment in non-Python environments,
+which are frequently employed for inference deployments. AOTInductor plays a pivotal role in the
+`export <https://pytorch.org/docs/main/export.html>`__ path, offering a means to execute an exported
+model independently of a Python runtime.
```
Comment (msaroufim) [inline docs/source/torch.compiler_aot_inductor.rst:18]: can make it clearer that this is for on server applications otherwise executorch pitch is similar

File: docs/source/torch.compiler_aot_inductor.rst [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+---------------------------
+
+Using AOTInductor, you can still author the model in Python. The following example demonstrates how to
+invoke ``aot_compile`` to transform the model into a shared library.
+
+.. note::
+
+   To execute the following code, it's essential to have a CUDA-enabled device on your machine.
+   If you do not possess a GPU, you can simply omit the ``.to(device="cuda")`` code within the snippet
+   below. In such a case, the script will compile the model code into a shared library that is optmized
```
Comment (msaroufim) [inline docs/source/torch.compiler_aot_inductor.rst:34]:
```suggestion
   below. In such a case, the script will compile the model code into a shared library that is optimized
```

File: docs/source/torch.compiler_aot_inductor.rst [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+            x = self.fc2(x)
+            x = self.sigmoid(x)
+            return x
+
+    with torch.no_grad():
+        m = Model().to("cuda")
+        example_inputs=(torch.randn(8, 10).to("cuda"),)
+        batch_dim = torch.export.Dim("batch", min=1, max=1024)
+        dynamic_shapes = {"x": {0: batch_dim}}
+        so_path = torch._export.aot_compile(m, example_inputs, dynamic_shapes=dynamic_shapes)
```
Comment (msaroufim) [inline docs/source/torch.compiler_aot_inductor.rst:61]: any plans for a non `_` API

File: docs/source/torch.compiler_aot_inductor.rst [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+            std::cerr << "Error: Unable to open model_so_path.txt." << std::endl;
+            return 1;
+        }
+        std::string model_so;
+        if (!std::getline(path_file, model_so)) {
+            std::cerr << "Error: File is empty." << std::endl;
+        }
+        path_file.close();
+
+        torch::inductor::AOTIModelRunnerCuda runner(model_so.c_str());
```
Comment (msaroufim) [inline docs/source/torch.compiler_aot_inductor.rst:112]: Add a comment on top of this line to highlight it since it's the most relevant part

File: docs/source/torch.compiler_aot_inductor.rst [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+
+    aot_inductor_example/
+        CMakeLists.txt
+        inference.cpp
+        model.py
+
+
+.. code-block:: shell
+
+    (nightly) [ ~/local/aot_inductor_example]$ mkdir build
```
Comment (msaroufim) [inline docs/source/torch.compiler_aot_inductor.rst:161]: i think you can skip these logs

Comment (msaroufim) [inline docs/source/torch.compiler_aot_inductor.rst:161]: i think you can skip these logs

File: docs/source/torch.compiler_aot_inductor.rst [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+    0.4611
+    0.4744
+    0.4811
+    0.4938
+    0.4193
+    [ CUDAFloatType{8,1} ]
+    Result from second inference:
+    0.4883
+    0.4703
+    [ CUDAFloatType{2,1} ]
```
Comment (msaroufim) [inline docs/source/torch.compiler_aot_inductor.rst:230]: do you wanna end by linking to the python wrapper as well?

**Review by msaroufim — APPROVED** [review]
minor nits

File: docs/source/torch.compiler_aot_inductor.rst (hunk 3 above)
Comment (desertfire) [reply docs/source/torch.compiler_aot_inductor.rst:61]: Not sure what the plan is for that, cc @angelayi @gmagogsfm

File: docs/source/torch.compiler_aot_inductor.rst (hunk 6 above)
Comment (desertfire) [reply docs/source/torch.compiler_aot_inductor.rst:230]: By python wrapper, do you mean loading the compiled shared library using pybind?

File: docs/source/torch.compiler_aot_inductor.rst [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+Model Compilation
+---------------------------
+
+Using AOTInductor, you can still author the model in Python. The following example demonstrates how to
+invoke ``aot_compile`` to transform the model into a shared library.
+
+.. note::
+
+   To execute the following code, it's essential to have a CUDA-enabled device on your machine.
```
Comment (albanD) [inline docs/source/torch.compiler_aot_inductor.rst:30]: nit: you can use `device = "cuda" if torch.cuda.is_available() else "cpu"`.

File: docs/source/torch.compiler_aot_inductor.rst [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+        def forward(self, x):
+            x = self.fc1(x)
+            x = self.relu(x)
+            x = self.fc2(x)
+            x = self.sigmoid(x)
+            return x
+
+    with torch.no_grad():
+        m = Model().to("cuda")
+        example_inputs=(torch.randn(8, 10).to("cuda"),)
```
Comment (albanD) [inline docs/source/torch.compiler_aot_inductor.rst:56]:
```suggestion
        example_inputs=(torch.randn(8, 10, device="cuda"),)
```

File: docs/source/torch.compiler_aot_inductor.rst (hunk 3 above)
Comment (angelayi) [reply docs/source/torch.compiler_aot_inductor.rst:59]: Sure, I think the API is stable enough now. I can make it public.

File: docs/source/torch.compiler_aot_inductor.rst [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+.. code-block:: cpp
+
+    #include <fstream>
+    #include <iostream>
+    #include <vector>
+    #include <torch/torch.h>
+    #include <torch/csrc/inductor/aoti_model_runner_cuda.h>
+
+    int main() {
+        torch::NoGradGuard no_grad;
```
Comment (albanD) [inline docs/source/torch.compiler_aot_inductor.rst:97]: You most likely want inference mode here and above.

File: docs/source/torch.compiler_aot_inductor.rst [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+            x = self.fc2(x)
+            x = self.sigmoid(x)
+            return x
+
+    with torch.no_grad():
+        m = Model().to("cuda")
+        example_inputs=(torch.randn(8, 10).to("cuda"),)
+        batch_dim = torch.export.Dim("batch", min=1, max=1024)
+        dynamic_shapes = {"x": {0: batch_dim}}
+        so_path = torch._export.aot_compile(m, example_inputs, dynamic_shapes=dynamic_shapes)
```
Comment (albanD) [inline docs/source/torch.compiler_aot_inductor.rst:59]: Feels weird to get a path back? Why is that not an input to specify the path?

File: docs/source/torch.compiler_aot_inductor.rst [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+
+    print(f"Compiled model into: {so_path}")
+    with open("model_so_path.txt", "w") as file:
```
Comment (albanD) [inline docs/source/torch.compiler_aot_inductor.rst:62]: having a path hardcoded that points to a file that contains a path where you should look feels a bit convoluted? It was not what I expected when reading the code quickly.

File: docs/source/torch.compiler_aot_inductor.rst [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+        }
+        std::string model_so;
+        if (!std::getline(path_file, model_so)) {
+            std::cerr << "Error: File is empty." << std::endl;
+        }
+        path_file.close();
+
+        // AOTIModelRunnerCuda dlopens the compiled shared library
+        torch::inductor::AOTIModelRunnerCuda runner(model_so.c_str());
+        std::vector<torch::Tensor> inputs = {torch::randn({8, 10}).to(at::kCUDA)};
```
Comment (albanD) [inline docs/source/torch.compiler_aot_inductor.rst:112]:
```suggestion
        std::vector<torch::Tensor> inputs = {torch::randn({8, 10}, at::kCUDA};
```
Same below

File: docs/source/torch.compiler_aot_inductor.rst [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+
+    aot_inductor_example/
+        CMakeLists.txt
+        inference.cpp
+        model.py
+
+
+.. code-block:: shell
+
+    (nightly) [ ~/local/aot_inductor_example]$ mkdir build
```
Comment (albanD) [inline docs/source/torch.compiler_aot_inductor.rst:160]: nit: feel free to remove the env and path headers.

File: docs/source/torch.compiler_aot_inductor.rst [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+        path_file.close();
+
+        // AOTIModelRunnerCuda dlopens the compiled shared library
+        torch::inductor::AOTIModelRunnerCuda runner(model_so.c_str());
+        std::vector<torch::Tensor> inputs = {torch::randn({8, 10}).to(at::kCUDA)};
+        std::vector<torch::Tensor> outputs = runner.run(inputs);
+        std::cout << "Result from first inference:"<< std::endl;
+        std::cout << outputs[0] << std::endl;
+
+        std::cout << "Result from second inference:"<< std::endl;
```
Comment (albanD) [inline docs/source/torch.compiler_aot_inductor.rst:117]: This is important to showcase the dynamic batch dimension mentioned above. It might be worth mentioning it again here.

File: docs/source/torch.compiler_aot_inductor.rst (hunk 10 above)
Comment (desertfire) [reply docs/source/torch.compiler_aot_inductor.rst:59]: Specifying an input can be done through additional options. I was debating with myself whether to do that here, mostly for the reason you are commenting later for using a file to pass the path between Python and C++. It will be some kind of hard-coded either way, but maybe you are right that specifying a path here feels more natural. I will change it.

Comment (albanD) [reply docs/source/torch.compiler_aot_inductor.rst:59]: tbh without reading the variable names, what I though happened here is that your functions returns a custom python object representing the binary and the open/write below was actually writing the .so.

I realized while backtracking that the open was not writing the .so haha

File: docs/source/torch.compiler_aot_inductor.rst (hunk 9 above)
Comment (bdhirsh) [reply docs/source/torch.compiler_aot_inductor.rst:97]: Pretty much all of the problems around inference mode had to do with the jit setting where we are in the middle of tracing / running our compile stack under inference mode.

But it should be 100% fine to turn inference mode here, where it will only be at runtime (and this is exactly the case where you'll see speedups from inference_mode over no_grad, if the generated inductor code here is using the pytorch dispatcher)

File: docs/source/torch.compiler_aot_inductor.rst (hunk 14 above)
Comment (bdhirsh) [reply docs/source/torch.compiler_aot_inductor.rst:117]: Out of curosity: If the user forgot to designate their batch dim as dynamic during export (so the generated kernel specialized on batch size of the example input), what would happen when we try to run inference here? Would we raise an error?

File: docs/source/torch.compiler_aot_inductor.rst (hunk 9 above)
Comment (desertfire) [reply docs/source/torch.compiler_aot_inductor.rst]: I changed the C++ code to use `c10::InferenceMode`, but left the Python code to use `torch.no_grad()` just to be safe. @bdhirsh , do you happen to have a reproducible example for the problem we were discussing earlier? In the minimum, we should track the problem with an issue if not fixing it immediately.

File: docs/source/torch.compiler_aot_inductor.rst (hunk 14 above)
Comment (desertfire) [reply docs/source/torch.compiler_aot_inductor.rst]: Currently AOTInductor ignores generating all those assert_size_stride, but I will change it to generate some assertion for checking guarded shapes.

File: docs/source/torch.compiler_aot_inductor.rst (hunk 9 above)
Comment (bdhirsh) [reply docs/source/torch.compiler_aot_inductor.rst]: The last issue we hit was actually fixed: https://github.com/pytorch/pytorch/pull/112163/files#diff-48bd497885cd293ff4c37e382ec2f604b472adfe48d155da49db00ff99db54f2R771

The current state is basically

(1) Using inference_mode **during** tracing provides zero benefit of no_grad(), and is a bit fragile because of the way it interacts with the dispatcher. I think we've fixed all existing issues I know of, but I wouldn't be surprised if more pop up. We should just make this work in all cases in the long term though, and keep fixing bugs as they come up.

(2) Using inference_mode at runtime is strictly better than no_grad

(3) With `torch.compile`, we end up executing both the tracing and runtime all in one go, so people inevitably run the tracing bits of torch.compile under inference_mode

(4) with export though, there's a clean barrier between the user API for exporting the model and running the model. So it seems fine in the tutorial to use `no_grad()` during export - in the long term we'd like it to be functionally identical to exporting under `inference_mode()`, and in the medium term we know it's a bit less likely for users to run into issues

File: docs/source/torch.compiler_aot_inductor.rst [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+        CMakeLists.txt
+        inference.cpp
+        model.py
+
+
+.. code-block:: shell
+
+    $ mkdir build
+    $ cd build
+    $ CMAKE_PREFIX_PATH=/home/$USER/local/miniconda3/envs/nightly/lib/python3.10/site-packages/torch/share/cmake cmake ..
```
Comment (albanD) [inline docs/source/torch.compiler_aot_inductor.rst:157]: nit: remove the CMAKE_PREFIX_PATH.  Users should have this configured on their end and this path is specific to your setup

File: docs/source/torch.compiler_aot_inductor.rst (hunk 6 above)
Comment (albanD) [reply docs/source/torch.compiler_aot_inductor.rst:179]: I would agree that the best way to reduce confusion here is to list all options and explain when to use which.
But maybe as a one-paragraph intro at the top.

**Review by albanD — APPROVED** [review]
I would add at least a sentence clarifying the existence of the python driver and a small pointer if users want to learn more + saying that they should use either one depending on their "driver code"?

File: docs/source/torch.compiler_aot_inductor.rst (hunk 15 above)
Comment (desertfire) [reply docs/source/torch.compiler_aot_inductor.rst:157]: I actually think keeping it would serve as an example. When I read https://pytorch.org/tutorials/advanced/cpp_frontend.html for the first time, I found `-DCMAKE_PREFIX_PATH=/path/to/libtorch` used in the example there very confusing, because I couldn't get it to work by specifying the path pointing to my libtorch (maybe that piece of information is simply out-of-date).

Comment (albanD) [reply docs/source/torch.compiler_aot_inductor.rst:157]: Can you make it abstract then: `/path/to/python/install/site-packages/torch/share/cmake` ?

**Review by albanD — APPROVED** [review]
